Emergency Access Button Cooldown

AlinaStarkova opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->
Cyberboss said I could add this
## About The Pull Request
recently on Terry some insufferable twat decided to spam emergency access (both AI and two others actually), which is not only text spam but SOUND spam too, it makes ears bleed
![image](https://user-images.githubusercontent.com/40489693/131564566-5c94dd46-d3ab-45a1-8090-4798bba8e0ed.png)


so now theres a cooldown
you have 3 uses to click the button without a cooldown, afterwhich you get a 30 second cooldown, after the 30 seconds are up, you can hit the button 3 more times without cooldown... alternatively every 30 seconds of not clicking the button resets to 3 clicks without a cooldown, so if you click once, wait 30 seconds, you can hit 3 times free still.

this cooldown shouldnt ever occur during a round unless someones spamming it


all tested, if you dont use up all 3 uses and wait 30 seconds you get your uses back as well, so you always have the ability to spam it a few times before the cooldown kicks in, and once CD ends you get all uses back as well
![image](https://user-images.githubusercontent.com/40489693/132009019-6b3e12d6-30ff-4d25-bdfc-b334768b1333.png)


<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Because abusing things which spam chat with massive bold words as well as makes a loud sound shouldnt be allowed

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Nari Harimoto
fix: Emergency Access can no longer be spammed to cause sound and text pain
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

AlinaStarkova committed:
Emergency Access Button Cooldown

ZephyrTFA commented on code/game/machinery/computer/communications.dm:
```diff
@@ -41,6 +41,12 @@
+	var/toggle_cooldown = 3 //the minimum cooldown of toggling the emergency access IN SECONDS
+	var/toggle_max_uses = 6 //how many uses can you toggle emergency access with before cooldowns start occuring BOTH ENABLE/DISABLE
+	var/time_last_toggled //when emergency access was last toggled
+	var/time_toggle_reset = 30 //how long one must wait without toggling to reset toggle_uses IN SECONDS
```
For all of these move the comment above it and have it be ```/// description``` this is for autodoc purposes

AlinaStarkova committed:
auto-doc?

idk what autodoc is but someone told me to write comments like this for autodoc

MrStonedOne commented:
> IF you hit it again within cooldown, nothing happens AND your cooldown DOUBLES to 6, so on, and so forth,

This is cancerous. stackoverflow used to do the same thing I can link to the million articles that explain why resetting cooldown periods when they haven't been expired is bad and harmful and user hostile.

AlinaStarkova commented:
> 
> 
> > IF you hit it again within cooldown, nothing happens AND your cooldown DOUBLES to 6, so on, and so forth,
> 
> This is cancerous. stackoverflow used to do the same thing I can link to the million articles that explain why resetting cooldown periods when they haven't been expired is bad and harmful and user hostile.

what? literally all this is saying is "30 seconds is the max cooldown" else you might accidently cooldown so hard that you can have to wait minutes, or if spammed, hours, i intended this to be a progressive punishment to a wouldbe spammer without fucking over everyone, else a spammer can spam the button to get a long cooldown, forcing everyone to have to wait to use it for a legitimate purpose

AlinaStarkova commented on code/game/machinery/computer/communications.dm:
```diff
@@ -365,6 +378,19 @@
+	++toggle_uses //add a use
+	if((time_last_toggled + time_toggle_reset SECONDS) < world.time) // after you havent pushed this for awhile, reset the cooldown and uses
+		toggle_uses = 1 //we just used this so we reset to 1, not 0
+		toggle_cooldown = 3 //original cooldown, change this to whatever you change the var to be
```
no it shouldnt, the var is 3 initially, the seconds is declared when in actual use, as per Cyberboss's request (i wanted to declare the seconds in the var itself but he said he didnt like that as much)

Ghilker commented on code/game/machinery/computer/communications.dm:
```diff
@@ -365,6 +378,19 @@
+		toggle_cooldown = 3 //original cooldown, change this to whatever you change the var to be
+	if(toggle_uses >= toggle_max_uses)
+		if((time_last_toggled + toggle_cooldown SECONDS) > world.time) //if were still in cooldown, let you know, and double the cooldown
+			toggle_cooldown *= 2
```
nvm this

AlinaStarkova replied:
thats what the 30 second timer for reset is for, watch the end of the video to see it done

MrStonedOne commented:
i mean increasing the cooldown, (or resetting it to start over again) is bad and user hostile.

AlinaStarkova commented on code/game/machinery/computer/communications.dm:
```diff
@@ -41,6 +41,17 @@
+	///when emergency access was last toggled
+	var/time_last_toggled
+	///how long one must wait without toggling to reset toggle_uses IN SECONDS
+	var/time_toggle_reset = 30
```
no, this is exactly as intended, its 30 seconds, seconds is declared in the proc, as per Cyberboss's request

AlinaStarkova replied:
doesnt matter, cyberboss wanted "SECONDS" declared in the proc, not in the var

AlinaStarkova commented:
> 
> 
> i mean increasing the cooldown, (or resetting it to start over again) because they used it before the cooldown ended is bad and user hostile.

i dont see how thats bad, this would literally ONLY occur if spammed, remember: your first 5 clicks is COMPLETLEY free, and as long as you dont click for 30 seconds your clicks are reset, so you will always be able to click whenever you want unless you click 5 times within 30 seconds, which..... would only ever be done by a potential spammer

MrStonedOne commented:
you don't get it

MrStonedOne commented:
doing all of that is fine.

doing it *when the countdown hasn't ended* is bad

AlinaStarkova commented:
> 
> 
> doing all of that is fine.
> 
> doing it _when the countdown hasn't ended_ is bad

no no, i totally get it, the issue i had was i cant track that stuff outside of when a button is clicked, since that code can only check the time on each click, i know theres a SSticker or whatever system that is always ticking to manage timers, but whatever, Cyberboss just said in discord that doing that would be overcomplicated to stop spammers, so he said just change it to a set cooldown instead

AlinaStarkova committed:
changed to static cooldown

AlinaStarkova commented:
making a dynamic cooldown that doesnt check on click is too complicated for me, speaking with cyberboss and MSO in discord ive decided to make a static cooldown with 3 free uses without cooldown

Mothblocks commented on code/game/machinery/computer/communications.dm:
```diff
@@ -365,6 +376,23 @@
+		to_chat(user, span_warning("This was your third free use without cooldown, you will not be able to use this again for [toggle_cooldown_seconds] seconds."))
+
+	if(toggle_uses > toggle_max_uses)
+		if((time_last_toggled + toggle_cooldown_seconds SECONDS) >= world.time)
```
Use the cooldown defines.

Mothblocks commented on code/game/machinery/computer/communications.dm:
```diff
@@ -365,6 +376,23 @@
+			to_chat(user, span_warning("Emergency Access is still in Cooldown! Once cooldown ends you will get [toggle_max_uses] free uses without cooldown again."))
+			return TRUE //dont use the button, we are in cooldown
+
+	if((time_last_toggled + toggle_cooldown_seconds SECONDS) < world.time) // after you havent pushed this for awhile, reset the uses
```
Same here.

Mothblocks commented on code/game/machinery/computer/communications.dm:
```diff
@@ -365,6 +376,23 @@
 
+/obj/machinery/computer/communications/proc/emergency_access_cooldown(mob/user)
+	if(toggle_uses == toggle_max_uses)
+		to_chat(user, span_warning("This was your third free use without cooldown, you will not be able to use this again for [toggle_cooldown_seconds] seconds."))
```
```suggestion
		to_chat(user, span_warning("This was your final free use without cooldown, you will not be able to trigger emergency access again for [toggle_cooldown_seconds] seconds."))
```

This means we can change the number 3 without having to update the text too, plus some minor changes.

Mothblocks commented on code/game/machinery/computer/communications.dm:
```diff
@@ -365,6 +376,23 @@
+
+	if(toggle_uses > toggle_max_uses)
+		if((time_last_toggled + toggle_cooldown_seconds SECONDS) >= world.time)
+			to_chat(user, span_warning("Emergency Access is still in Cooldown! Once cooldown ends you will get [toggle_max_uses] free uses without cooldown again."))
```
```suggestion
			to_chat(user, span_warning("Emergency access is still on cooldown."))
```

The other information isn't necessary unless the player plans on spamming it again, IMO.

AlinaStarkova committed:
uses cooldown defines

Mothblocks commented on code/game/machinery/computer/communications.dm:
```diff
@@ -365,6 +376,24 @@
+		var/time_left = DisplayTimeText(COOLDOWN_TIMELEFT(src, emergency_access_cooldown), 1)
+		to_chat(user, span_warning("Emergency Access is still in cooldown for [time_left]!"))
+		return TRUE //dont use the button, we are in cooldown
+	else if((last_toggled + EMERGENCY_ACCESS_COOLDOWN) < world.time)
```
This should be a cooldown too, right?

AlinaStarkova replied:
no, because its not cooling down, its just checking if you havent touched it in 30 seconds, you can still use it during that time, if i did a cooldown then i would have to initiate a cooldown every 30 seconds nonstop for this part when thats not what i want

Mothblocks commented on code/game/machinery/computer/communications.dm:
```diff
@@ -365,6 +376,24 @@
 			minor_announce("Due to staff shortages, your station has been approved for delivery of access codes to secure the Captain's Spare ID. Delivery via drop pod at [get_area(pod_location)]. ETA 120 seconds.")
 
+/obj/machinery/computer/communications/proc/emergency_access_cooldown(mob/user)
+	if(toggle_uses == toggle_max_uses) //you have used up 3 free uses already, do it one more time and start a cooldown
+		to_chat(user, span_warning("This was your third free use without cooldown, you will not be able to use this again for [DisplayTimeText(EMERGENCY_ACCESS_COOLDOWN)]."))
```
```suggestion
	if(toggle_uses == toggle_max_uses) //you have used up free uses already, do it one more time and start a cooldown
		to_chat(user, span_warning("This was your last free use without cooldown, you will not be able to use this again for [DisplayTimeText(EMERGENCY_ACCESS_COOLDOWN)]."))
```

```suggestion
		to_chat(user, span_warning("This was your last free use without cooldown, you will not be able to use this again for [DisplayTimeText(EMERGENCY_ACCESS_COOLDOWN)]."))
```

I still don't want the text to be hardcoded in case we want to change the number easier.

AlinaStarkova committed:
Update code/game/machinery/computer/communications.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>
